crates/shared/src/event_handling.rs

```diff
@@ -161,24 +263,162 @@ where
+            .to_block((*block_range.end()).into())
             .block_page_size(500)
             .query_paginated()
             .await?
             .map_err(Error::from))
     }
+
+    fn update_last_handled_blocks(&mut self, blocks: &[BlockNumberHash]) {
+        tracing::debug!(
+            "blocks to update into last_handled_blocks: {:?} - {:?}, last_handled_blocks: {:?} - {:?}",
+            blocks.first(),
+            blocks.last(),
+            self.last_handled_blocks.first(),
+            self.last_handled_blocks.last(),
+        );
+        if blocks.is_empty() {
+            return;
+        }
+        // delete forked blocks
+        self.last_handled_blocks
+            .retain(|block| block.0 < blocks.first().unwrap().0);
+        // append new canonical blocks
+        self.last_handled_blocks.extend(blocks.iter());
+        // cap number of blocks to MAX_REORG_BLOCK_COUNT
+        let start_index = self
+            .last_handled_blocks
+            .len()
+            .saturating_sub(MAX_REORG_BLOCK_COUNT as usize);
+        self.last_handled_blocks = self.last_handled_blocks[start_index..].to_vec();
+        tracing::debug!(
+            "last_handled_blocks after update: {:?} - {:?}",
+            self.last_handled_blocks.first(),
+            self.last_handled_blocks.last(),
+        );
+    }
+}
+
+/// Try to shorten the latest_blocks by detecting the reorg from previous event update.
+/// If no reorg can be detected (for example, when `handled_blocks` is shorter then the
+/// reorg depth) then fallback to full `latest_blocks` as a safe measure.
+fn detect_reorg_path<'a>(
+    handled_blocks: &[BlockNumberHash],
+    latest_blocks: &'a [BlockNumberHash],
+) -> (&'a [BlockNumberHash], bool) {
+    // in most cases, latest_blocks = handled_blocks + 1 newest block
+    // therefore, is it more efficient to put the handled_blocks in outer loop,
+    // so everything finishes in only two iterations.
+    for handled_block in handled_blocks.iter().rev() {
+        for (i, latest_block) in latest_blocks.iter().enumerate().rev() {
+            if latest_block == handled_block {
+                // found the same block in both lists, now we know the common ancestor, don't include the ancestor
+                let is_reorg = handled_block != handled_blocks.last().unwrap();
+                return (&latest_blocks[i + 1..], is_reorg);
+            }
+        }
+    }
+
+    // reorg deeper than the EventHandler history (`handled_blocks`), return full list
+    let is_reorg = !handled_blocks.is_empty();
+    (latest_blocks, is_reorg)
+}
+
+/// Splits range into two disjuctive consecutive ranges, second one containing last (up to)
+/// MAX_BLOCKS_QUERIED elements, first one containing the rest (if any)
+fn split_range(
+    range: RangeInclusive<u64>,
+) -> Result<(Option<RangeInclusive<u64>>, RangeInclusive<u64>)> {
+    ensure!(
+        MAX_BLOCKS_QUERIED > 0,
+        "MAX_BLOCKS_QUERIED must be greater than zero"
+    );
```

Comment:
The only way this function can fail is through this ensure. But the thing we're asserting is entirely static so it feels weird. I'd rather do an `assert` and have the function never fail.

Reply:
Changed to `assert` but then clippy doesn't like `assert(true)`. Ended up removing it completely, since I don't see a case where would someone set `MAX_REORG_BLOCK_COUNT` to 0.